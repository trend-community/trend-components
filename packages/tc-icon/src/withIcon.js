import React from 'react';
import PropTypes from 'prop-types';

const defaultStyle = {
  display: 'inline-block',
  fill: 'currentColor',
  verticalAlign: 'middle',
  height: 'auto',
  width: 'inherit'
};

function withIcon(WrappedComponent) {
  function WithIcon({ size, unit, style, ...rest }) {
    const styles = {
      ...defaultStyle,
      ...style
    };

    return <WrappedComponent
      {...rest}
      height={`${parseFloat(size)}${unit}`}
      width={`${parseFloat(size)}${unit}`}
      style={styles} />;
  }

  WithIcon.propTypes = {
    size: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    unit: PropTypes.oneOf(['em', 'rem']),
    style: PropTypes.object
  };
  WithIcon.defaultProps = {
    size: 1,
    unit: 'em'
  };
  WithIcon.displayName = `WithIcon(${getDisplayName(WrappedComponent)})`;

  return WithIcon;
}

function getDisplayName(WrappedComponent) {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component';
}

export default withIcon;
