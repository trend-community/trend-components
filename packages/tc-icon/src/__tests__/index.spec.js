import React from 'react';
import { shallow } from 'enzyme';

import withIcon from '../';

const IconComponent = props => <svg {...props} />;

const Icon = withIcon(IconComponent);

describe('tc-icon (withIcon)', () => {
  it('should set the wrapped icon with the correct props.', () => {
    const wrapper = shallow(<Icon />);
    const WrappedComponent = wrapper.find('IconComponent').dive();
    expect(WrappedComponent.props()).toMatchSnapshot();
  });

  it('should allow the size to be updated.', () => {
    const wrapper = shallow(<Icon size={2} />);
    const WrappedComponent = wrapper.find('IconComponent').dive();
    expect(WrappedComponent.props()).toMatchSnapshot();
  });

  it('should merge the styles.', () => {
    const style = {
      overflow: 'hidden',
      height: '1rem',
      width: '1rem'
    };
    const wrapper = shallow(<Icon style={style} />);
    const WrappedComponent = wrapper.find('IconComponent').dive();
    expect(WrappedComponent.prop('style')).toMatchSnapshot();
  });
});
