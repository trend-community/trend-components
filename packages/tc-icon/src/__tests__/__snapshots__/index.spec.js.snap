// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tc-icon (withIcon) should allow the size to be updated. 1`] = `
Object {
  "height": "2rem",
  "style": Object {
    "display": "inline-block",
    "fill": "currentColor",
    "height": "auto",
    "verticalAlign": "middle",
    "width": "inherit",
  },
  "width": "2rem",
}
`;

exports[`tc-icon (withIcon) should merge the styles. 1`] = `
Object {
  "display": "inline-block",
  "fill": "currentColor",
  "height": "1rem",
  "overflow": "hidden",
  "verticalAlign": "middle",
  "width": "1rem",
}
`;

exports[`tc-icon (withIcon) should set the wrapped icon with the correct props. 1`] = `
Object {
  "height": "1rem",
  "style": Object {
    "display": "inline-block",
    "fill": "currentColor",
    "height": "auto",
    "verticalAlign": "middle",
    "width": "inherit",
  },
  "width": "1rem",
}
`;
